### ISSUE TITLE
Finding the most recent rows matching a filter

### ISSUE THREAD

**Issue by HashFail** [Stratio/cassandra-lucene-index#318]
I hope this is the appropriate place for this. I searched for some sort of forum for this but couldn't find one, and I saw several other support questions here, so here goes: 

I'm trying to figure out how best to leverage this plugin for my specific use case. I have an application where I'm inserting data at a very high rate. At the same time, I need to run searches on the same table of the form: 

Find the X most recent rows with timestamp <= Y where field foo = bar. 

However, most of the time, "Y" in this scenario will be the current timestamp, and when it's not it'll be close to it, and X is very small (25-100 out of millions of rows). I'm wondering how/if I can take advantage of this fact. My present setup is to use an index as follows: 

```
CREATE CUSTOM INDEX tweets_index ON tweets (lucene)
USING 'com.stratio.cassandra.lucene.Index'
WITH OPTIONS = {
   'refresh_seconds': '1',
   'schema': '{
      fields: {
         username: {type: "string"},
         content: {type: "text"}, 
	 created: {type: "integer"},
         parent_tweet: {type: "string"}
      }
   }'
};
```

"created" is a unix timestamp (integer). One thing I'm wondering about is if there's any value in using a timestamp. 

My queries are of the form

```
SELECT * FROM tweets WHERE lucene = '{
            filter : {
                type : "boolean",
                must: [
                    {
                        type : "range",
                        field : "created",
                        lower : 0,
                        upper : * upper timestamp *
                    }
                    * other filters go here *
                ]
            },
            sort  : { fields: [ { field:"created", reverse:true } ] }
        }' limit * limit goes here*;
```

Under this current setup, indexing is pretty fast, with the average time to insert (with network latency) being around 29ms. However, searching takes much longer, averaging 415ms. This is much higher than I'd like (I'm aiming to have the 95th and 99th percentiles be ~200ms and ~600ms). 

What I am wondering is, is there some better way I can be indexing/searching this data? Perhaps by using TimeUUIDs and token ranges or something? I imagine most of the slowdown is happening because my searches use a "sort," and thus are hitting every node in the cluster. Or maybe there is some way I can indicate to that this data is "temporal", and recently inserted rows should be cached? Or am I using the plugin correctly and the problem is perhaps elsewhere in my system? 

Apologies again if this is not the correct place to ask for this sort of advice. Thank you for your time.

**Comment by ealonsodb** [comment]
Hi @HashFail: 

Could you please show us your table creation statement? It depends on the specific use case.  I think you need to test what is correct in your case. 

Your test case seem just like a twitter news feed.  Is it common that the tweeted resulted by that search belongs to just one day(today) ? 

If you partition your table with the timestamp day you can restrict your search to that day and the search will be much faster. 
In the other hand with this approach you will have writing and searching hotspots (Writes and reads  for today will ask to the same cassandra node). 

Now, each index will only have data related to one day. Searches will be much faster. Also, if you need to search in more than one day you can use the [IN](http://docs.datastax.com/en/cql/3.3/cql/cql_using/useQueryIN.html) keyword. 

You can play with the partition key granularity. (day, hour, minute, second) to get your required performance.

Hope this helps

**Comment by HashFail** [comment]
Hi @ealonsodb, 

Yes, that is the application I'm referring to. My table creation statement is as follows: 

```
create table tweets(
  tweet_id timeuuid, 
  created int,
  content varchar, 
  retweet timeuuid,
  username varchar,
  parent_tweet timeuuid,
  lucene text,
  media set<varchar>,
  primary key(tweet_id)
);
```
I considered an approach similar to the one you described. One of my concerns was that it creates more overhead in the application code, potentially requiring more than one query if there aren't enough tweets for the given partition key. However, that could just be the SQL programmer in me; I'm learning the adage that you should let the datastore do most of the work doesn't always apply to Cassandra. Now that I know this is an accepted/good practice in Cassandra, I'll attempt it. 

However, the issue you raise, read/write hotspots, does concern me, since I have disabled virtual nodes, meaning hotspots will disproportionately affect physical shards, potentially having a serious impact on performance. I disabled virtual nodes because I was experiencing performance problems with them with this plugin (something mentioned in the documentation). 

Anyway, thank you for your advice! This is a side project of mine to evaluate Cassandra and this plugin, so I may not be able to provide updates immediately, but I'll post my results when I can. Thank you again!

**Comment by ealonsodb** [comment]
Hi @HashFail: 

What i mean with [IN keyword](http://docs.datastax.com/en/cql/3.3/cql/cql_using/useQueryIN.html) is that you can get more-than-one partition rows 

```
CREATE TABLE tweets_by_day (
  day text, 
  tweet_id timeuuid, 
  created int,
  content varchar, 
  lucene text,
  ...
  PRIMARY KEY((tweet_id, day))
);

INSERT INTO tweets_by_day(tweet_id, day. created, content) VALUES(now(),'2017-03-02',now(),'Bla bla bla');
INSERT INTO tweets_by_day(tweet_id, day. created, content) VALUES(now(),'2017-03-03',now(),'Bla bla bla');
INSERT INTO tweets_by_day(tweet_id, day. created, content) VALUES(now(),'2017-03-04',now(),'Bla bla bla');
INSERT INTO tweets_by_day(tweet_id, day. created, content) VALUES(now(),'2017-03-05',now(),'Bla bla bla');


SELECT * FROM tweets_by_day WHERE day IN ('2017-03-02', '2017-03-03', '2017-03-04', '2017-03-05') AND lucene = '{...}' LIMIT 100;
```

Hotspots concern everyone. Disabling virtual tokens does not prevent you to have hotspots. What i mean with hotspots is: 

If you have a partition key related with time (for example day) and you insert rows today , those inserts will arrive in the same cassandra node (or one replica). If you search for results for today you are asking the same cassandra node group for the data. With this approach you are not using the entire cluster(only one node group at every day) 

Hope this helps

**Comment by HashFail** [comment]
Hi @ealonsodb, 

Sorry for the delay in responding. Yes, I understood what you meant. I went ahead and used an approach as you described using a granularity of one second, and I achieved much better results. 

My table creation now looks like this: 
```
create table tweets(
  tweet_id timeuuid, 
  created int,
  content varchar, 
  retweet timeuuid,
  username varchar,
  parent_tweet timeuuid,
  lucene text,
  media set<varchar>,
  reply boolean,
  primary key(created, tweet_id)
);
```

And queries look like this:

`SELECT * FROM tweets WHERE created AND lucene = '{}'`

Allow filtering is I only use lucene when I need to perform a text search. If I need to filter by other attributes, I use 'ALLOW FILTERING'. I'm going to experiment with indexing them in luecne as well, but I haven't seen severe penalties with allow filtering so far, and most searches do not require it anyway. 

So the initial results have been very good, however. I added more nodes to my cluster, which improved performance, but before I made the changes you suggested I was getting ~6ms for inserting, ~270ms for search. After the changes, both take around 12ms. 

Thank you again for your insight!
